plugin/plugin.go

```diff
@@ -12,28 +20,102 @@ import (
+}
+type PluginSource struct {
+	Impl Source
+}
+
+func (p *PluginSource) Server(mb *hplugin.MuxBroker) (interface{}, error) {
+	return &pluginSourceRPCServer{Impl: p.Impl, mb: mb}, nil
 }
 
-// Plugin is the principal interface that all plugins must implement.
-type Plugin interface {
-	// NewSource creates a new Source given a configuration.
-	// If the plugin does not support acting as a source, then it should return ErrNotImplemented.
-	NewSource(ctx context.Context, config map[string]interface{}) (Source, error)
-	// NewDestination creates a new Destination given a configuration.
-	// If the plugin does not support acting as a destination, then it should return ErrNotImplemented.
-	NewDestination(ctx context.Context, config map[string]interface{}) (Destination, error)
+func (p *PluginSource) Client(mb *hplugin.MuxBroker, c *rpc.Client) (interface{}, error) {
+	return &pluginSourceRPC{client: c, mb: mb}, nil
+}
+
+type PluginDestination struct {
+	Impl Destination
+}
+
+func (p *PluginDestination) Client(mb *hplugin.MuxBroker, c *rpc.Client) (interface{}, error) {
+	return &pluginDestinationRPC{client: c, mb: mb}, nil
+}
+
+func (p *PluginDestination) Server(mb *hplugin.MuxBroker) (interface{}, error) {
+	return &pluginDestinationRPCServer{Impl: p.Impl, mb: mb}, nil
+}
+
+func NewPlugin(ctx context.Context, path string) (Destination, Source, error) {
+	handshakeConfig := hplugin.HandshakeConfig{
+		ProtocolVersion:  PluginMagicProtocalVersion,
+		MagicCookieKey:   PluginMagicCookieKey,
+		MagicCookieValue: PluginCookieValue,
+	}
+
+	// pluginMap is the map of plugins we can dispense.
+	pluginMap := map[string]hplugin.Plugin{
+		"destination": &PluginDestination{},
+		"source":      &PluginSource{},
+	}
+
+	logger := hclog.New(&hclog.LoggerOptions{
+		Name:       "plugin",
+		JSONFormat: true,
+		Output:     os.Stdout,
+		Level:      hclog.Debug,
+	})
+
+	// We're a host! Start by launching the plugin process.
+	client := hplugin.NewClient(&hplugin.ClientConfig{
+		HandshakeConfig: handshakeConfig,
+		Plugins:         pluginMap,
+		Cmd:             exec.CommandContext(ctx, path),
+		Logger:          logger,
+	})
+
+	// Connect via RPC
+	rpcClient, err := client.Client()
+	if err != nil {
+		client.Kill()
+		return nil, nil, err
+	}
+	go func() {
+		// TODO find out how to exit gracefully
+		// Sometime the RPC server keeps running usings lots of CPU.
+
+		// EDIT: We can also remove this because we kill the process when using exec.CommandContext
+		<-ctx.Done()
+		logger.Info("closing rpc client")
+		if err := rpcClient.Close(); err != nil {
+			logger.Error("failed to close rpc client", "err", err.Error())
+		}
+		client.Kill()
+	}()
+
+	mErr := &multierror.Error{}
+	rawD, errD := rpcClient.Dispense("destination")
+	if errD != nil {
+		mErr = multierror.Append(mErr, errD)
+	}
+	rawS, errS := rpcClient.Dispense("source")
+	if errS != nil {
+		mErr = multierror.Append(mErr, errS)
+	}
+	if mErr.Len() == 2 {
+		rpcClient.Close()
+		client.Kill()
```

Comment:
if we only get 1 error that's ok then? we don't have to kill the client?

Reply:
yes maybe we get an error because only the source is implemented. In that case we want to be happy with the destination.